### ISSUE TITLE
fast show for Nat

### ISSUE THREAD

**Issue by mechvel** [agda/agda-stdlib#678]
Matthew has suggested to discuss certain current questions here, on GitHub.

>> ``toDecimalChars : ℕ → List Char   (via _%_, _div_)``,

> A function very much like this already exists in Data.Digit. Have you seen it?

1) Data.Digit.toDigits  is "exponentially" slow, and  Data.Nat.Show.show  is via  toDigits
For example,  ``p = 2 ^ 30`` is evaluated fast, because _*_ is via a built-in.
Then, ``(Data.Nat.Show.show p)`` has less than 10 digits to print,   and this takes 51 sec on a 3 GHz machine.
And  ``show' `` made via ``toDecimalChars`` takes  0.002 sec.

2) Still  I cannot re-implement   toDigits  via  ``_%_, _div_``  so far.
Because toDigits returns List (Fin base) and a certain proof. And conversion to (Fin base) needs to apply ``fromℕ≤`` to the inequality   ``r<base``   related to result of   (n % base). At this point the witness   ``r<base`` is being inspected, and it explodes.
There is needed an investigation of how to improve the part of ``fromℕ≤``.

3) So that I suggest, so far, to add  ``toDecimalChars`` to standard and to re-implement  Data.Nat.Show.show via  ``toDecimalChars``.

< Certain code shown,  it has been converted to the pull request >

**Comment by MatthewDaggitt** [comment]
This sounds like a good plan. I'd open a PR to add `toDecimalChars` in `Data.Nat.Show`.

**Comment by mechvel** [comment]
Probably, the below code needs to be put into   Data/Nat/Show.agda :

```
open import Induction.WellFounded using (Acc)
open Acc using (acc)

<-acc :  {n : ℕ} → Acc _<_ n     -- This means that _<_ is well-founded on ℕ
                                                 -- ("each number in ℕ is accessible from 0 by _<_").
<-acc {n} = Induction.Nat.<-wellFounded n

open ≤-Reasoning using () renaming (begin_ to ≤begin_; _∎ to _≤end;
                                                            _≡⟨_⟩_ to _≡≤[_]_; _≤⟨_⟩_ to _≤[_]_)
1* = *-identityˡ
*1 = *-identityʳ 

-- Conversion from unary to decimal system.
toDigitNats : (base : ℕ) → base > 1 → ℕ → List ℕ
toDigitNats 0                  ()
toDigitNats 1                  1>1        _ =  contradiction 1>1 (<-irrefl refl)
toDigitNats (suc (suc b)) base>1 x =  aux x <-acc [] 
  where
  base = suc (suc b);   pbase = pred base

  aux : (n : ℕ) → Acc _<_ n → List ℕ → List ℕ 
  aux 0          _           =  (0 ∷_)
  aux (suc n) (acc wf) =
    let
      n' = suc n;     q = n' div base;     r = n' % base 

      n'≡r+q*base :  n' ≡ r + q * base
      n'≡r+q*base =  a≡a%n+[a/n]*n n' pbase
    in
    case q ≟ 0
    of \
    { (yes _)   → (r ∷_)
    ; (no q≢0) → -- use  n' ≡ r + q * 10 
        let
          pq            = pred q
          suc-pq≡q  = m≢0⇒suc[pred[m]]≡m q≢0
          q<q*base  =
                ≤begin suc q                      ≡≤[ cong suc (sym suc-pq≡q) ]
                           suc (suc pq)            ≡≤[ cong suc (sym (*1 (suc pq))) ]
                           suc ((suc pq) * 1)      ≤[ *-monoʳ-< pq base>1 ]
                           (suc pq) * base       ≡≤[ cong (_* base) suc-pq≡q ]
                           q * base
                ≤end

          q<n' = ≤begin suc q                 ≤[ q<q*base ]
                                 q * base            ≤[ m≤m+n (q * base) r ] 
                                 q * base + r     ≡≤[ +-comm _ r ]
                                 r + q * base     ≡≤[ sym n'≡r+q*base ]
                                 n'
                     ≤end
        in
        aux q (wf _ q<n') ∘ (r ∷_)
    }

toDigitChar :  (n : ℕ) → Char
toDigitChar n =   Char.fromNat (n + 48) 

1<10 : 1 < 10        
1<10 = s≤s (s≤s z≤n)

toDecimalChars : ℕ → LChar
toDecimalChars =
               map toDigitChar ∘ toDigitNats 10 1<10

show :    ℕ → String                                     -- to replace  Dat.Nat.Show.show
show =   Str.fromList ∘ toDecimalChars

-- test : show 1090 ≡ "1090"
-- test = refl
```

**Comment by MatthewDaggitt** [comment]
Yup, something like that code. The next step is to open a pull request (PR). 

However before you actually add the code to `Data.Nat.Show` please can you read the [style guide](https://github.com/agda/agda-stdlib/blob/master/notes/style-guide.md) first and make sure your code obeys the conventions described within.

**Comment by mechvel** [comment]
@style-guide:
> The list of module import should be in alphabetical order.

1) The list of the imported module names _or_  the list of items in ``used (...)`` ?
2) Experimental lib version of April 1, 2019 does not follow any of this. For example, see Data.List.Base.agda.
3) Is not it more reasonable to put ahead those ``import <Module> `` lines which modules import less of other modules? (so that the modules containing more complex notions are named later).

**Comment by MatthewDaggitt** [comment]
> The list of the imported module names or the list of items in used (...) ?

The list of imported module names

> Experimental lib version of April 1, 2019 does not follow any of this. For example, see `Data.List.Base`.

It's an aspirational style guide. Most of the files obey it, and whenever we alter an older file we try to make it conform. You're right `Data.List.Base` doesn't obey this rule at the moment because it's an older file that hasn't been changed much since we created the style guide. When I next get round to updating it, I'll put the module imports in order. For instance `Data.List.Properties` does obey it.

> Is not it more reasonable to put ahead those import <Module> lines which modules import less of other modules? (so that the modules containing more complex notions are named later).

Which module is more complex is a judgement call which varies from person to person. Alphabetical order allows a user to quickly look up whether or not a given module is used.

**Comment by mechvel** [comment]
Matthew,
I managed to do this pull request -- with a great help from Jesper and Apostolis.
I would write down the process, and hope that would be able to do further PR, if any.
Ii is on 
``https://github.com/mechvel/agda-stdlib/tree/experimental``

It is  Data.Nat.Show.agda,   the "show" function implementation changed.
I have modified the source to follow  the  style guide  most closely.
--------
Sergei
